# 15. Исключения

[Книга по Java](https://vk.com/@codeblog-dzhoshua-bloh-java-effektivnoe-programmirovanie) - читать главу об исключениях.

Что нужно знать об исключениях?

1) Структуру исключения
2) Иерархия исключений
3) Какие исключения где встречаются
4) Создание собственных классов-исключений
5) Обработка исключений
6) Best practices по исключениям
7) Пять ключевых слов для работы с исключениями (try, catch, throw, throws, finally).


* Ошибки

- Ошибки, возникающие при нарушении синтаксиса языка программирования - ошибки времени компиляции. Виноват программист.

- Ошибки, возникающие при неправильной эксплуатации ПО (пользователь - дурак), либо, возникающие при неправильно написанной логике программы. - ошибки времени выполнения. Виноват программист.

## 15.00 Исключения в Java

* Рассматриваем только ошибки (иключения) времени выполнения.

* Каждое исключения в Java - это отдельный объект, созданный на основе какого-либо класса, именно поэтому с исключениями можно работать.

## 15.01 Иерархия исключений в Java

* `IllegalArgumentException` - класс исключений, который предназначен как раз для ситуаций, когда программист хочет выбросить исключение, на, например, неверно введенные данные. Потомок `RuntimeException`.

* `ArithmeticException` - класс исключений, возникающих при вычислительных ошибках в программе. Потомок `RuntimeException`.

* `NullPointerException` - класс исключений, возникающих при попытке вызова каких-либо методов у объектных переменных, не имеющих ссылки на конкретный объект. Потомок `RuntimeException`.

* `StackOverflowError` - класс исключений, связанных с превышением размера стека вызовов для функций и процедур (слишком много вызовов). Потомок `VirtualMachineError`.

* `OutOfMemoryError` - класс исключений, связанных с превышением доступной памяти. Потомок `VirtualMachineError`.

* `VirtualMachineError` - класс исключений, связанных с перебоями виртуальной машины. Помоток `Error`.

* `Error` - класс исключений, связанных с перебоями работы системы. Потомок `Throwable`

* `RuntimeException` - потомок `Exception`.

* `FileNotFoundException` - класс исключений, связанных с обращениями к несуществующим файлам. Потомок `IOException`.

* `IOException` - класс исключений, связанных с вводом-выводом. Потомок `Exception`.

* `Exception` - потомок `Throwable`.

* `Throwable` - класс-предок всех исключительных ситуаций. 

Самый большой интерес представляет структура этого класса:

```JAVA
class Throwable {
	...
	// детальное описание ошибки
	// Java heap space
	// / by zero
	// файл которого нет (Не удается найти указанный файл)
	private String detailMessage;

	// массив данных о том, что именно привело к ошибке
	private StackTraceElement[] stackTrace;

	private Throwable cause = this;
}

```

* `StackTraceElement`

```JAVA
public final class StackTraceElement implements java.io.Serializable {
    // Normally initialized by VM (public constructor added in 1.5)
    private String declaringClass;
    private String methodName;
    private String fileName;
    private int lineNumber;
```

## 15.02 Обработка исключений в Java

* Обработка исключения - это перехват исключения, и выполнение определенного кода в такой нештатной ситуации.

```JAVA
try {
	// блок кода, в котором потенциально может возникнуть исключение
} catch (Класс-исключения-которое-хотим-перехватить объектная-переменная-исключения) {
	// блок кода, который выполняется при возникновении исключения
}
```

* Перехватывать можно любые исключения, то есть все потомки `Throwable`.

* Если в блоке `try` находится код, который потенциально может сгенерировать разные типы исключений, нужно писать несколько `catch` для каждого типа исключения.

* Если в блоке `catch` указано какое-либо исключение, то в данный блок попадают не только объекты этого исключения, но и его потомки.

* Если у исключения несколько потомков, но для каких-то из них следует предусмотреть отдельную обработку, то нужно написать эти исключения выше. Потому что если выше будет указан предок, то Java не сможет правильно передать исключение в `catch`, потому что общее стоит раньше частного.

Не правильно:

```JAVA
catch (Exception e) {
    System.out.println("Я на Exception");
	e.printStackTrace();
} catch (FileNotFoundException e) {
	System.out.println("Файл мы не нашли");
}
```

Правильно:

```JAVA
catch (FileNotFoundException e) {
	System.out.println("Файл мы не нашли");
} catch (Exception e) {
	System.out.println("Я на Exception");
	e.printStackTrace();
}
```

## 15.03 Выброс исключений

* Необходим, когда мы хотим остановить работу программы с определенной ошибкой там, где Джавой она не предусмотрена.

* Используется ключевое слово `throw`:

```JAVA
throw new КлассИсключения();
```

## 15.04 Создание пользовательских исключений

* Пользовательские исключения позволяют более конкретно описать суть проблемы.

* Позволяют программе отслеживать специфичные исключения.

* При созданиии своего исключения необходимо наследоваться от `RuntimeException`. На самом деле, можно наследоваться от любого класса-потомока `Throwable`, но не надо.

## 15.05 Группы исключений

* Проверяемые исключения (`checked exceptions`) - `Throwable`, `Exception` и его потомки, кроме `RuntimeException` и его потомков. Данные исключения обязаны быть обработанными в явном виде. Иначе программа не скомпилируется. Если метод/конструктор выбрасывает проверяемое исключение, в его сигнатуре должен быть указан оператор `throws`.

* Непроверяемый исключения (`unchecked exceptions`) - `RuntimeException` и его потомки, `Error` и его потомки.

* Если метод вызывает другой метод, который выбрасивает проверяемое исключение - у вас два варианта - либо отработать с этим исключением тут же, либо перебросить его наверх с помощью оператора `throws`, данный оператор работает аналогично `catch` в плане иерархии ошибок.

## 15.06 Правильная работа с исключениями

1. Никогда не создавать своих проверяемых исключений, всегда наследуемся от `RuntimeException`

2. Никогда не обрабатываем ошибки так, как советует IntelliJ IDEA. Потому что программа в этом случае продолжает свое выполнение:

```JAVA
try {
            inputStream = new FileInputStream("input.txt");
        } catch (FileNotFoundException e) {
        	// здесь вы ничего с ошибкой не сделали, вы просто ее проигнорировали, при этом не остановили выполнение программы
            e.printStackTrace();
        }
```

3. Когда наталкиваетесь на проверяемое исключение, вы оборачиваете ее в непроверяемое. Таким образом вы останавливаете выполнение программы и при этом не перекладываете ответсвенность далее.

4. Никогда не использовать `throws` если вам лень обрабатывать ошибки. Ошибки необходимо обрабатывать на месте. Потому что таким образом, все программисты могут пробросить ошибку вплоть до страницы клиента.